src/main/scala/com/codacy/duplication/pmd/Cpd.scala

```diff
@@ -0,0 +1,154 @@
+import codacy.docker.api.duplication._
+import codacy.docker.api.{DuplicationConfiguration, Source}
+import com.codacy.api.dtos.{Language, Languages}
+import net.sourceforge._
+import net.sourceforge.pmd.cpd
+import net.sourceforge.pmd.cpd.Match
+import play.api.libs.json.{JsNumber, JsValue}
+
+import scala.collection.JavaConverters._
+import scala.util.{Failure, Success, Try}
+
+object Cpd extends DuplicationTool {
+
+  private val allLanguages: List[Language] =
+    List[Language](
+      Languages.Python,
+      Languages.Ruby,
+      Languages.Java,
+      Languages.Javascript,
+      Languages.Scala,
+      Languages.CSharp)
+
+  override def apply(
+    path: Source.Directory,
+    language: Option[Language],
+    options: Map[DuplicationConfiguration.Key, DuplicationConfiguration.Value]): Try[List[DuplicationClone]] = {
+    val baos = new ByteArrayOutputStream()
+    val stdErr = System.err
+
+    System.setErr(new PrintStream(baos, true, "utf-8"))
+    val directoryPath: Path = (File.currentWorkingDirectory / path.path).path
+
+    getLanguages(language).map { languages =>
+      runCPD(directoryPath, languages, options)(stdErr)
+    } match {
+      case Success(matches) => Success(matches)
+      case Failure(e) =>
+        val errString = new String(baos.toByteArray, StandardCharsets.UTF_8)
+        val msg =
+          s"""|Failed to execute duplication: ${e.getMessage}
+              |std:
+              |$errString
+         """.stripMargin
+
+        Failure(new Exception(msg, e))
+    }
+  }
+
+  private def runCPD(directory: Path,
+                     languages: List[(pmd.cpd.Language, Int)],
+                     options: Map[DuplicationConfiguration.Key, DuplicationConfiguration.Value])(
+    errStream: PrintStream): List[DuplicationClone] = {
+
+    languages.flatMap { lang =>
+      val configuration = getConfiguration(lang, options)
+      runWithConfiguration(configuration, directory)(errStream)
+    }
+
+  }
+
+  private def runWithConfiguration(config: pmd.cpd.CPDConfiguration, directory: Path)(
+    errStream: PrintStream): List[DuplicationClone] = {
+    val cpd = new pmd.cpd.CPD(config)
+    cpd.addRecursively(directory.toFile)
+
+    System.setErr(errStream)
+
+    cpd.go()
+
+    val res: util.Iterator[Match] = cpd.getMatches
+    res.asScala.map(matchToClone(_, directory)).toList
+  }
+
+  private def getLanguages(language: Option[Language]): Try[List[(pmd.cpd.Language, Int)]] = {
+
+    val languagesOpt: Option[Try[List[(cpd.Language, Int)]]] =
+      language.map { lang =>
+        convertLanguageToConfig(lang) match {
+          case Some(config) => Success(List(config))
+          case None =>
+            Failure(new Throwable(s"Cpd.getLanguages: there is no language with key $lang"))
+        }
+      }
+
+    languagesOpt.getOrElse(Success(allLanguages.flatMap(convertLanguageToConfig)))
+  }
+
+  private def convertLanguageToConfig(language: Language): Option[(pmd.cpd.Language, Int)] = {
```

Comment:
This is named "convertLanguageToConfig" which returns still a language, along with an int, which then is used by another function named "convertLanguageToConfig" that takes this as an input and then returns a configuration.

So this does not return a configuration but instead some data that is needed to create the configuration, hence this is a wrong name.